### DISCUSSION TITLE
How to build extensions with additional dependencies?

### DISCUSSION THREAD

**Discussion by Eliophot** [directus/directus#10735] [Archive]
Hello,
I have to make an endpoint with additional dependencies. I had them in package.json file of the extension. But the build inevitably does not find the new dependencies (require).
Should I also add them in the package.json file of the api ?
Anothers questions : 
* the --watch functionnality don't work with additional files.
* the path (directus:extension) just build the index.js. How can build my additional files ?

**Comment by rijkvanzanten** [comment]
Is this in regards to the API extensions, or app extensions? The app extensions are bundled, so all dependencies should be included. As for the API extensions, you might be right regarding the package.json of the root of the project, but i'll defer to @nickrum for more information

**Reply by Eliophot** [reply]
It's for the API extensions and yes, that's what I did. It's not easy when you have several extensions, but it's ok, I managed with concurrently + nodemon + copy-files-from-to + npm-watch for now.

**Reply by nickrum** [reply]
Alternatively, you could also use the extensions-sdk to bundle your API extension to a single file. This will pull in any dependency you are using without having to add them to `package.json`. The only downside is that the generated `index.js` file will become quite large.

**Reply by ched-dev** [reply]
I noticed when I used an `index.js` entry point it did not include the files imported with `require()` statements, causing the output to be invalid. If I switched to an `index.ts` it would copy the imported code and build correctly. 

Should I be using rollup plugins to include all the contents into one file or does the `directus-extension build` command support it somehow?

**Reply by bibatjed** [reply]
> I noticed when I used an index.js entry point it did not include the files imported with require() statements, causing the output to be invalid.

Any updates on this one? I also encountered this issue.

**Reply by bibatjed** [reply]
> > I noticed when I used an index.js entry point it did not include the files imported with require() statements, causing the output to be invalid.
> 
> Any updates on this one? I also encountered this issue.

using esm fixed the issue.

**Comment by nickrum** [comment]
> the --watch functionnality don't work with additional files.

This sounds like a bug to me. The extensions-sdk should watch any file you import from the extension's main entrypoint. Mind opening an issue?

> the path (directus:extension) just build the index.js. How can build my additional files ?

This works as intended. The extension entrypoint imported by directus should ideally always be a single file. What exactly are you trying to achieve using additional files?

**Reply by Eliophot** [reply]
I use Directus as a proxy with other APIs. One of these API is in soap and so a lot of xml.
The index.js file was too big, so I had to create these files for more clarity in my code...

**Reply by rokdd** [reply]
I can confirm that `the --watch functionnality don't work with additional files.` . And also it does not reload these files, need to restart directus. And in the build command it does not copy the file to the dist folder.
